File: cli/src/voidrift_cli/doctor.py
"""Diagnostic checks for voidrift doctor (REQ-U-16, REQ-U-16b)."""

from __future__ import annotations

import shutil
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

CheckResult = Literal["pass", "warn", "fail"]


@dataclass
class Check:
    name: str
    result: CheckResult
    message: str = ""
    fix_hint: str = ""
    auto_fixable: bool = False


def run_checks(fix: bool = False) -> list[Check]:
    """Run all doctor checks. If fix=True, auto-remediate where safe."""
    from .config import _voidrift_home, load_config, get_models_file

    checks: list[Check] = []
    home = _voidrift_home()

    # 1. Config file
    cfg_path = home / "config.yml"
    if not cfg_path.exists():
        checks.append(Check("config", "fail", f"Config not found: {cfg_path}", "Run 'make sync' to create it"))
    else:
        try:
            load_config()
            checks.append(Check("config", "pass", str(cfg_path)))
        except Exception as e:
            checks.append(Check("config", "fail", f"Config parse error: {e}"))

    # 2. Models file
    try:
        mf = get_models_file()
        if not mf.exists():
            checks.append(Check("models", "fail", f"Models file not found: {mf}", "Configure models_file in config.yml"))
        else:
            import yaml
            try:
                data = yaml.safe_load(mf.read_text()) or {}
                models = data.get("models", [])
                count = len(models) if isinstance(models, list) else 0
                checks.append(Check("models", "pass", f"{mf} ({count} models)"))
            except yaml.YAMLError as e:
                checks.append(Check("models", "fail", f"Models YAML error: {e}"))
    except Exception:
        checks.append(Check("models", "warn", "Could not determine models file path"))

    # 3. Skills
    skills_dir = home / "resources" / "skills"
    if skills_dir.is_dir():
        files = list(skills_dir.glob("*.md"))
        bad = []
        for f in files:
            try:
                text = f.read_text(encoding="utf-8")[:512]
                if not text.startswith("---"):
                    bad.append(f.name)
            except OSError:
                bad.append(f.name)
        if bad:
            msg = f"{len(files)} files, {len(bad)} missing frontmatter"
            checks.append(Check("skills", "warn", msg, "Add YAML frontmatter to skill files"))
        else:
            checks.append(Check("skills", "pass", f"{len(files)} files, all parseable"))
    else:
        checks.append(Check("skills", "warn", "Skills directory not found", "Run 'make sync'"))

    # 4. Log directory
    log_dir = home / "logs"
    if log_dir.is_dir():
        try:
            test_file = log_dir / ".doctor-test"
            test_file.write_text("ok")
            test_file.unlink()
            checks.append(Check("logs", "pass", f"{log_dir} — writable"))
        except OSError:
            checks.append(Check("logs", "fail", f"{log_dir} — not writable"))
    else:
        if fix:
            log_dir.mkdir(parents=True, exist_ok=True)
            checks.append(Check("logs", "pass", f"Created {log_dir}"))
        else:
            checks.append(Check("logs", "fail", f"{log_dir} missing", "Run 'voidrift doctor --fix'", auto_fixable=True))

    # 5. Disk space
    try:
        usage = shutil.disk_usage(Path.cwd())
        free_gb = usage.free / (1024 ** 3)
        if free_gb < 0.1:
            checks.append(Check("disk", "fail", f"{free_gb:.1f}GB available", "Free disk space"))
        elif free_gb < 1.0:
            checks.append(Check("disk", "warn", f"{free_gb:.1f}GB available"))
        else:
            checks.append(Check("disk", "pass", f"{free_gb:.0f}GB available"))
    except OSError:
        checks.append(Check("disk", "warn", "Could not check disk space"))

    # 6. Model entry validation (REQ-U-16b)
    try:
        mf = get_models_file()
        if mf.exists():
            checks.extend(_check_model_entries(mf))
    except Exception:
        pass

    # 7. Optional tool dependencies (REQ-U-16a)
    checks.extend(_check_optional_deps())

    return checks


_VALID_PROTOCOLS: frozenset[str] = frozenset({"openai", "anthropic"})


def _check_model_entries(models_file_path: Path) -> list[Check]:
    """Validate individual model entries for required fields and valid protocol (REQ-U-16b)."""
    import yaml

    try:
        data = yaml.safe_load(models_file_path.read_text()) or {}
    except Exception:
        return []  # Already reported by the models file YAML check

    defaults = data.get("defaults", {}) or {}
    models = data.get("models", {}) or {}
    if not isinstance(models, dict):
        return []

    failures: list[str] = []
    for alias, entry in models.items():
        if not isinstance(entry, dict):
            failures.append(f"'{alias}': not a valid entry")
            continue
        m = {**defaults, **entry}
        for field in ("base_url", "api_key", "model_id"):
            if field not in m:
                failures.append(f"'{alias}': missing required field '{field}'")
        protocol = m.get("protocol", "openai")
        if protocol not in _VALID_PROTOCOLS:
            failures.append(
                f"'{alias}': invalid protocol '{protocol}' "
                f"(valid: {', '.join(sorted(_VALID_PROTOCOLS))})"
            )

    if failures:
        return [Check(
            "model-entries",
            "fail",
            "; ".join(failures),
            "Fix missing/invalid fields in models.yml then re-run 'voidrift doctor'",
        )]
    count = len(models)
    return [Check("model-entries", "pass", f"{count} entr{'y' if count == 1 else 'ies'} valid")]


_OPTIONAL_DEPS: list[tuple[str, str, str]] = [
    ("tree_sitter", "tree-sitter", "code_analysis tool"),
    ("tree_sitter_python", "tree-sitter-python", "code_analysis for Python"),
    ("fitz", "pymupdf", "read_document for PDF"),
    ("docx", "python-docx", "read_document for DOCX"),
    ("openpyxl", "openpyxl", "read_document for XLSX"),
]


def _check_optional_deps() -> list[Check]:
    """Check availability of optional tool dependencies."""
    import importlib

    checks: list[Check] = []
    for module_name, pip_name, purpose in _OPTIONAL_DEPS:
        try:
            importlib.import_module(module_name)
            checks.append(Check(f"optional:{pip_name}", "pass", f"{pip_name} — available ({purpose})"))
        except (ImportError, ModuleNotFoundError):
            checks.append(Check(
                f"optional:{pip_name}", "warn",
                f"{pip_name} not installed ({purpose})",
                f"pip install {pip_name}",
            ))
    return checks

File: src/commands/doctor.ts
/**
 * Doctor command: diagnostic checks (REQ-U-16).
 */

import { existsSync, readFileSync, mkdirSync, accessSync, constants, statfsSync } from "node:fs";
import { join } from "node:path";
import { parse as parseYaml } from "yaml";
import { voidriftHome, getModelsFile } from "../config.js";

export interface CheckResult {
  name: string;
  status: "pass" | "warn" | "fail";
  message: string;
}

const VALID_PROTOCOLS = new Set(["openai", "anthropic"]);

export function runChecks(fix = false): CheckResult[] {
  const results: CheckResult[] = [];
  const home = voidriftHome();
  let configOk = false;

  // Config file
  const configPath = join(home, "config.yml");
  if (!existsSync(configPath)) {
    results.push({ name: "config", status: "warn", message: `${configPath} not found` });
  } else {
    try {
      parseYaml(readFileSync(configPath, "utf-8"));
      results.push({ name: "config", status: "pass", message: "config.yml valid" });
      configOk = true;
    } catch (e: unknown) {
      results.push({ name: "config", status: "fail", message: `config.yml parse error: ${e instanceof Error ? e.message : e}` });
    }
  }

  // Models file — skip if config is broken
  if (configOk || !existsSync(configPath)) {
    let modelsPath: string;
    try { modelsPath = getModelsFile(); } catch { modelsPath = ""; }
    if (!modelsPath || !existsSync(modelsPath)) {
      results.push({ name: "models_file", status: "fail", message: `models file not found` });
    } else {
      try {
        const data = parseYaml(readFileSync(modelsPath, "utf-8")) ?? {};
        results.push({ name: "models_file", status: "pass", message: "models file valid" });
        const models = (data.models ?? {}) as Record<string, Record<string, unknown>>;
        let allValid = true;
        for (const [alias, entry] of Object.entries(models)) {
          for (const field of ["base_url", "api_key", "model_id"]) {
            if (!(field in entry)) {
              results.push({ name: "model_entry", status: "fail", message: `Model '${alias}' missing '${field}'` });
              allValid = false;
            }
          }
          const protocol = entry.protocol ? String(entry.protocol) : "openai";
          if (!VALID_PROTOCOLS.has(protocol)) {
            results.push({ name: "model_entry", status: "fail", message: `Model '${alias}' invalid protocol '${protocol}'` });
            allValid = false;
          }
        }
        if (allValid && Object.keys(models).length) {
          results.push({ name: "model_entries", status: "pass", message: `${Object.keys(models).length} model(s) valid` });
        }
      } catch (e) {
        results.push({ name: "models_file", status: "fail", message: `models file parse error: ${e}` });
      }
    }
  }

  // Log directory
  const logDir = join(home, "logs");
  if (!existsSync(logDir)) {
    if (fix) {
      mkdirSync(logDir, { recursive: true });
      results.push({ name: "log_dir", status: "pass", message: "log directory created" });
    } else {
      results.push({ name: "log_dir", status: "warn", message: `${logDir} missing. Run with --fix to create.` });
    }
  } else {
    try {
      accessSync(logDir, constants.W_OK);
      results.push({ name: "log_dir", status: "pass", message: "log directory writable" });
    } catch {
      results.push({ name: "log_dir", status: "fail", message: "log directory not writable" });
    }
  }

  // Disk space
  try {
    const stats = statfsSync(".");
    const availGb = (stats.bavail * stats.bsize) / (1024 ** 3);
    results.push({
      name: "disk_space",
      status: availGb < 0.1 ? "fail" : availGb < 1.0 ? "warn" : "pass",
      message: `${availGb.toFixed(1)} GB available`,
    });
  } catch {
    results.push({ name: "disk_space", status: "warn", message: "could not check disk space" });
  }

  // Skills directory
  const skillsDir = join(home, "resources", "skills");
  results.push(existsSync(skillsDir)
    ? { name: "skills", status: "pass", message: "skills directory present" }
    : { name: "skills", status: "warn", message: "skills directory not found" });

  return results;
}

export function formatResults(results: CheckResult[]): string {
  return results.map(r => {
    const icon = r.status === "pass" ? "✓" : r.status === "warn" ? "⚠" : "✗";
    return `  ${icon} ${r.name}: ${r.message}`;
  }).join("\n");
}
